### PR TITLE
Compile on numpy2

### PR DESCRIPTION
**Pull request by jeromekelleher**
Part of supporting numpy >= 2. Following the ABI [advice](https://numpy.org/doc/stable/dev/depending_on_numpy.html#numpy-2-abi-handling) from numpy, we should be able to compile against numpy version >= 2, and also have a runtime dependency < 2. Hopefully this is all we need to do initially to ship compatible wheels and conda packages (which are also broken)

### CODE REVIEW

**Comment by jeromekelleher** [comment]
Looks like this is mostly working now. Outstanding issues:

- Some tests are failing on most recent numpy 1.x version and also on numpy 2.x. 
- We need numpy 2.x compatible kastore distributions, so parking this until that's ready.

**Comment by jeromekelleher** [comment]
Shall I pick this up or is it simpler to start again @benjeffery? The kastore conda-forge packages should be built and available in the next few hours, so that should unblock us here.

**Comment by benjeffery** [comment]
> Shall I pick this up

I'm working on this now!

File: python/setup.py [hunk 1]
```diff
@@ -1,22 +1,18 @@
...
 from setuptools import setup
 from setuptools.command.build_ext import build_ext
 
-
 IS_WINDOWS = platform.system() == "Windows"
 
 
-# Obscure magic required to allow numpy be used as a 'setup_requires'.
-# Based on https://stackoverflow.com/questions/19919905
 class local_build_ext(build_ext):
```
Comment (jeromekelleher) [inline python/setup.py:11]: Do we need this at all any more?

File: python/tests/test_ld_matrix.py [hunk 2]
```diff
@@ -1034,7 +1034,7 @@ def pi2_unbiased(
         w_Ab = state[1, k]
         w_aB = state[2, k]
         w_ab = n - (w_AB + w_Ab + w_aB)
-        with suppress_division_by_zero_warning():
+        with np.errstate(over="ignore", divide="ignore", invalid="ignore"):
```
Comment (jeromekelleher) [inline python/tests/test_ld_matrix.py:1037]: Wouldn't it be better to update the original function here rather than duplicating?

**Review by jeromekelleher — COMMENTED** [review]
Actual code changes look small, great!

File: python/setup.py (hunk 1 above)
Comment (benjeffery) [reply python/setup.py]: Good spot, removed in 03e40911

File: python/tests/test_ld_matrix.py (hunk 2 above)
Comment (benjeffery) [reply python/tests/test_ld_matrix.py]: Fixed in a8cb721b

File: python/setup.py (hunk 1 above)
Comment (benjeffery) [reply python/setup.py]: Hmm, removing this worked locally, but not on CI.

**Comment by benjeffery** [comment]
All seems to be working - I'll squash and merge.

**Comment by benjeffery** [comment]
Realised we were doing wheels for 3.12, but not testing on it. Fixing that now

**Review by jeromekelleher — COMMENTED** [review]
LGTM! What's the strategy for pushing an updated release - cherry pick this back onto last release tag?

**Comment by benjeffery** [comment]
> LGTM! What's the strategy for pushing an updated release - cherry pick this back onto last release tag?

Seems the simplest approach.

**Pull request merged**